streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+      concat_txt_playlists.append(concat_txt_playlist)
+    
+    return concat_txt_playlists
+  
+  @staticmethod
+  def concat_aud_common(all_aud_playlists: List[List['MediaPlaylist']]
+                        ) -> Dict[AudioCodec,
+                                  Dict[str,
+                                       Dict[AudioChannelLayout,
+                                            List['MediaPlaylist']]]]:
+    """A common method that is used to divide the audio playlists into a structure
+    that is easy to process.
+    
+    This common method is used by `concat_aud` and `concat_aud_only` methods.
+    """
+    
+    # Extract all the codecs, languages, and channel layouts.
+    codecs: Set[AudioCodec] = set()
+    langs: Set[str] = set()
+    channels: Set[AudioChannelLayout] = set()
+    for aud_playlists in all_aud_playlists:
+      for aud_playlist in aud_playlists:
+        assert isinstance(aud_playlist.codec, AudioCodec)
+        codecs.add(aud_playlist.codec)
+        channels.add(aud_playlist.channel_layout)
+        langs.add(aud_playlist.language)
+    
+    # Create and initialize a division map.
+    division: Dict[AudioCodec,
+                   Dict[str,
+                        Dict[AudioChannelLayout,
+                             List['MediaPlaylist']]]] = {}
+    
+    for codec in codecs:
+      division[codec] = {}
+      for lang in langs:
+        division[codec][lang] = {}
+        for channel in channels:
+          division[codec][lang][channel] = []
+    
+    # This logic inside here is done on period basis.
+    for aud_playlists in all_aud_playlists:
+      # Initialize a mapping between audio codecs and language to a list of
+      # media playlists with channel layouts available.
+      codec_lang_division: Dict[AudioCodec, Dict[str, List[MediaPlaylist]]] = {}
+      for codec in codecs:
+        codec_lang_division[codec] = {}
+        for lang in langs:
+          codec_lang_division[codec][lang] = []
+      # For every audio playlist in this period, append it to the matching 
+      # codec/language.
+      for aud_playlist in aud_playlists:
+        assert isinstance(aud_playlist.codec, AudioCodec)
+        codec_lang_division[aud_playlist.codec][aud_playlist.language].append(
+            aud_playlist)
+      # Sort and replace the missing languages in the codec_lang_division map.
+      for codec in codecs:
+        for lang in langs:
+          # If this language for this codec in this period has no media playlists
+          # with a for any channel layout, this means that the language itself
+          # is missing.  We will try to find a substitution for it.
+          if not len(codec_lang_division[codec][lang]):
+            aud_playlist_options = [codec_lang_division[codec][lang][0] for
+                                    lang in langs
+                                    if len(codec_lang_division[codec][lang])]
+            sub_lang = MediaPlaylist._fit_missing_lang(aud_playlist_options,
```

Comment:
This substitution is not as big as the one you do in the player(based on channel count, bitrate, sample rate, etc..).
It is only based on the language only for simplicity, so if some language has no playlists at all with any channel layout, we query `_fit_missing_lang()` with the all the media playlists we have for that period and a language to fit upon, the method will return a best fit based on the language only.

Reply:
Sounds good.  You should be able to count on similar channels for the same reason as similar resolutions in video: everything went through the same pipeline config.